File: Data Structures & Algorithms/merge-two-sorted-linked-lists/submission-2.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def mergeTwoLists(self, list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:
        if not list1:
            return list2
        if not list2:
            return list1
        
        if list1.val > list2.val:
            return self.mergeTwoLists(list2, list1)
        
        c1 = list1
        c2 = list2
        prev = ListNode(0, None)
        while c1 and c2:
            if c1.val <= c2.val:
                prev = c1
                c1 = c1.next
            else:
                next2 = c2.next

                prev.next = c2
                c2.next = c1

                prev = c2
                c2 = next2
        if c2:
            prev.next = c2
        return list1

File: Data Structures & Algorithms/reverse-nodes-in-k-group/submission-2.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    ListNode* getKth(ListNode* cur, int k){
        while(cur && k > 0){
            cur = cur->next;
            k--;
        }
        return cur;
    }
    ListNode* reverseKGroup(ListNode* head, int k) {
        ListNode* res = new ListNode(0, head);
        ListNode* groupPrev = res;

        while(true){
            ListNode* kth = getKth(groupPrev, k);
            if(!kth) break;

            ListNode* groupNext = kth->next;
            ListNode* cur = groupPrev->next;
            ListNode* prev = groupNext;
            while(cur != groupNext){
                ListNode* tmp = cur->next;
                cur->next = prev;
                prev = cur;
                cur = tmp;
            }

            ListNode* tmp = groupPrev->next;
            groupPrev->next = kth;
            groupPrev = tmp;
        }
        return res->next;
    }
};
